### DISCUSSION TITLE
Common usuage for .all() for select * from tablname and select columnname from tablename

### DISCUSSION THREAD

**Discussion by MangalaUral** [sqlalchemy/sqlalchemy#11841] [Usage Questions]
Hi,

We are in the process of upgrading our SQLAlchemy version from 1.3 to 2.0. In version 1.3, we were using self.session.query. As part of the upgrade to version 2.0, we are transitioning to using select for queries and executing them with self.session.execute(self.current_query).all().

However, we’ve noticed some differences in the output format between the two versions for certain types of queries:

    For a query like:

SELECT columnname FROM tablename

    In version 1.3, the output was: (Decimal(1),)
    In version 2.0, the output is: (Decimal(1),)

For a query like:

    SELECT * FROM tablename

        In version 1.3, the output was: <Tablename object at 0x7428a7177e20>
        In version 2.0, the output is: (<Tablename object at 0x7428a7177e20>,)

The key difference is that in version 2.0, the output for SELECT * queries has changed to return a tuple, while in version 1.3 it directly returned the object.

Could you please advise on a common approach to handle both kinds of queries in SQLAlchemy 2.0?

Thank you!

**Comment by CaselIT** [comment]
Hi,

You can use scalars() if the query returns only a eingle entity.

Also note that the suggested path is to first go to 1.4 fix a warning there then upgrade to 2.0.

There are multiple pages on the docs on how upgrade, that would explain the changes, including this one

**Reply by MangalaUral** [reply]
Hi,
Thank you for the quick response.
Yes, we are first upgrading to 1.4 then we are moving 2.0.  It is not single value, the output would be something like this [(<Tablename object at 0x7428a7177e20>, <Tablename object at 0x7428a7177e20>, <Tablename object at 0x7428a7177e20>,)]. We do not want to change the way we are accessing the results also. So we need a suggestion which works for both types.

**Reply by CaselIT** [reply]
It's not clear what query are you trying to run. Can you clarify?

**Reply by MangalaUral** [reply]
I have a query like:

python

self.session.execute(select(db.Tablename)).all()

In SQLAlchemy 1.4, this returns results like:

[(<Tablename object at 0x7428a7177e20>), (<Tablename object at 0x7428a7177e20>), (<Tablename object at 0x7428a7177e20>,)]

Whereas in SQLAlchemy 1.3, the output was:


[<Tablename object at 0x7428a7177e20>, <Tablename object at 0x7428a7177e20>, <Tablename object at 0x7428a7177e20>]

In version 1.4, each result is wrapped in a tuple, while in version 1.3 the objects were returned directly. 
I need the result set of 1.4 to behave like 1.3. So what you would suggest for this ? Could you explain also why it is behaving differently ?

**Reply by FranckRodriguez** [reply]
Hello, i had the same issue and decided to change my queries form:

I use session.scalar when single object is expected
I use session.scalars when list of single objects is expected
I use session.execute when list of tuples (or list of list) of some variables is expected

For the "why", idk. My supervisors lied to me saying we were working on SQLALchemy 2.0. I saw i was under 1.4 several month later.
Only thing I know is that the 1.4 can andle with 2.0 queries format because of some automatic redirection

**Reply by CaselIT** [reply]
As mentioned use scalars if you want a list of scalar results.
If you want the rationale, again as mentioned, read the migration docs
